File: trees/minimum_depth_of_binary_tree_bfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0

        queue = deque([root])
        depth = 0

        while queue:
            n = len(queue)
            for _ in range(n):
                current_node = queue.popleft()
                if current_node.left is None and current_node.right is None:
                        return depth + 1
                if current_node.left:
                        queue.append(current_node.left)
                if current_node.right:
                        queue.append(current_node.right)
            depth += 1
        return depth

File: trees/minimum_depth_of_binary_tree_bfs.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def min_depth(root)
    return 0 if root.nil?
    queue = [root]
    depth = 0
    
    while !queue.empty?
        n = queue.size
        n.times do
            current_node = queue.shift
            if current_node.left.nil? && current_node.right.nil?
                return depth + 1
            end
            queue << current_node.left if current_node.left
            queue << current_node.right if current_node.right
        end
        depth += 1
    end
    depth
end